p2p/peersHolder/peersHolder.go

```diff
@@ -134,30 +165,49 @@ func (ph *peersHolder) Remove(peerID core.PeerID) {
 	shard, index, _ := ph.getShardAndIndexForPeer(peerID)
 	ph.removePeerFromMapAtIndex(shard, index)
 
-	connAddrSlice := pidData.connectionAddressSlice
+	connAddress := pidData.connectionAddress
 
 	delete(ph.peerIDs, peerID)
 
-	_, isPreferredPubKey := ph.connAddrSliceToPeerInfo[connAddrSlice]
-	if isPreferredPubKey {
-		// don't remove the entry because all the keys in this map refer to preferred connections and a reconnection might
-		// be done later
-		ph.connAddrSliceToPeerInfo[connAddrSlice] = nil
-	}
+	ph.removePeerInfoAtConnectionAddress(peerID, connAddress)
 
 	_, isWaitingForShardID := ph.tempPeerIDsWaitingForShard[peerID]
 	if isWaitingForShardID {
 		delete(ph.tempPeerIDsWaitingForShard, peerID)
 	}
 }
 
-// getKnownSlice checks if the connection address contains any of the initial preferred connection address slices
+// removePeerInfoAtConnectionAddress removes the entry associated with the provided pid from connAddrToPeersInfo map
+// it never removes the map key as it may be reused on a further reconnection
+func (ph *peersHolder) removePeerInfoAtConnectionAddress(peerID core.PeerID, connAddr string) {
+	peersInfo := ph.connAddrToPeersInfo[connAddr]
+	if peersInfo == nil {
+		return
+	}
+
+	var index int
+	var pInfo *peerInfo
+	for index, pInfo = range peersInfo {
+		if peerID == pInfo.pid {
+			break
+		}
+	}
+
+	peersInfo = append(peersInfo[:index], peersInfo[index+1:]...)
```

Comment:
if the peer does not exist, then you will remove index 0. can you treat this case?